File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // unordered_set<char>cset;
        // int maxi = INT_MIN;
        // int j = 0, i = 0;
        // while(i < s.size()){
        //     if(cset.find(s[i]) != cset.end()){
        //         cset.insert(s[i]);
        //         maxi = max(maxi, i - j + 1);
        //         i++;

        //     }else {
        //         cset.erase(s[j]);
        //         j++;
        //     }
        // }
        // return maxi;
        unordered_map<char, int>cmap;
        int j = 0;
        int maxi = 0;
        for(int i = 0; i<s.size();i++){
            if(cmap.find(s[i]) != cmap.end()) {
                j = max(j, cmap[s[i]] + 1); 
            }
            cmap[s[i]] = i;
            maxi = max(maxi, i - j + 1);
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map_char = defaultdict(int)
        left = 0
        res = 0
        for right in range(len(s)):
            if s[right] in map_char:
                left = max(left, map_char[s[right]] + 1)
            map_char[s[right]] = right
            res = max(res, right - left + 1)
        return res
        